File: research3/python/multi_thread.py
import math
import os
import sys
from concurrent.futures import ThreadPoolExecutor

UPPER_BOUND = 5_000_000


def is_prime(n: int) -> bool:
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, math.isqrt(n) + 1, 2):
        if n % i == 0:
            return False
    return True


def process_chunk(start: int, end: int) -> tuple[int, int]:
    """Return (count, sum) of primes in [start, end)."""
    count = 0
    total = 0
    for n in range(start, end):
        if is_prime(n):
            count += 1
            total += n
    return count, total


def main() -> None:
    num_cpus = os.cpu_count() or 1
    chunk_size = max(1, UPPER_BOUND // num_cpus)

    ranges = [
        (i, min(i + chunk_size, UPPER_BOUND + 1))
        for i in range(2, UPPER_BOUND + 1, chunk_size)
    ]

    with ThreadPoolExecutor(max_workers=num_cpus) as executor:
        results = list(executor.map(lambda r: process_chunk(*r), ranges))

    prime_count = sum(c for c, _ in results)
    prime_sum = sum(s for _, s in results)

    print(f"Upper bound: {UPPER_BOUND} | Cores: {num_cpus} | Primes found: {prime_count} | Checksum: {prime_sum}")


if __name__ == "__main__":
    main()

File: research3/rust/src/bin/multi_thread.rs
use rayon::prelude::*;

const UPPER_BOUND: u64 = 5_000_000;

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }
    let limit = (n as f64).sqrt() as u64 + 1;
    let mut i = 3;
    while i <= limit {
        if n % i == 0 {
            return false;
        }
        i += 2;
    }
    true
}

fn main() {
    let (prime_count, prime_sum): (u64, u64) = (2u64..=UPPER_BOUND)
        .into_par_iter()
        .filter(|&n| is_prime(n))
        .fold(
            || (0u64, 0u64),
            |(count, sum), n| (count + 1, sum + n),
        )
        .reduce(|| (0u64, 0u64), |(c1, s1), (c2, s2)| (c1 + c2, s1 + s2));

    println!(
        "Upper bound: {} | Cores: {} | Primes found: {} | Checksum: {}",
        UPPER_BOUND,
        rayon::current_num_threads(),
        prime_count,
        prime_sum
    );
}